src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.streaming.async;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4FastDecompressor;
+import org.apache.cassandra.exceptions.ChecksumMismatchException;
+import org.apache.cassandra.net.async.AppendingByteBufInputPlus;
+import org.apache.cassandra.utils.ChecksumType;
+
+/**
+ * A serialiazer for stream compressed files (see package-level documentation). Much like a typical compressed
+ * output stream, this class operates on buffers or chunks of the data at a a time. The format for each compressed
+ * chunk is as follows:
+ *
+ * - int - compressed payload length
+ * - int - uncompressed payload length
+ * - int - checksum of lengths
+ * - bytes - compressed payload
+ * - int - checksum of uncompressed data
+ *
+ * Note: instances are not thread-safe.
+ */
+class StreamCompressionSerializer
+{
+    /**
+     * Length of heaer data, which includes compressed length, uncompressed length, the their chunksum.
+     */
+    private static final int HEADER_LENGTH = 12;
+
+    enum DeserializeState { LENGTHS, PAYLOAD }
+
+    private final LZ4Compressor compressor;
+    private final LZ4FastDecompressor decompressor;
+
+    private DeserializeState deserializeState;
+    private ChunkLengths chunkLengths;
+    private final ByteBuffer lengthChecksumBuffer = ByteBuffer.allocate(8);
+
+    StreamCompressionSerializer(LZ4Compressor compressor, LZ4FastDecompressor decompressor)
+    {
+        this.compressor = compressor;
+        this.decompressor = decompressor;
+        deserializeState = DeserializeState.LENGTHS;
+    }
+
+    public ByteBuf serialize(ByteBuf in, int version) throws IOException
+    {
+        final ChecksumType checksum = ChecksumType.CRC32;
```

Comment:
So... this a weak spot for me. Because I don't remember the rationale for why 32-bits isn't enough or the math behind it. 

See page 13
https://users.ece.cmu.edu/~koopman/pubs/KoopmanCRCWebinar9May2012.pdf

I'm not actually sure if it keeps getting worse as code word length increases. I'll need to search some more.